gdnative-core/src/variant.rs

```diff
@@ -884,4 +929,177 @@ impl FromVariant for Variant {
+    }
+}
+
+impl<T: ToVariant> ToVariant for Option<T> {
+    fn to_variant(&self) -> Variant {
+        match &self {
+            Some(thing) => thing.to_variant(),
+            None => Variant::new(),
+        }
+    }
+}
+
+impl<T: FromVariant> FromVariant for Option<T> {
+    fn from_variant(variant: &Variant) -> Option<Self> {
+        if variant.is_nil() {
+            Some(None)
+        }
+        else {
+            T::from_variant(variant).map(Some)
+        }
+    }
+}
+
+/// Wrapper type around a `FromVariant` result that may not be a success
+#[derive(Clone, Debug)]
+pub struct MaybeNot<T>(Result<T, Variant>);
+
+impl<T: FromVariant> FromVariant for MaybeNot<T> {
+    fn from_variant(variant: &Variant) -> Option<Self> {
+        Some(MaybeNot(T::from_variant(variant).ok_or_else(|| variant.clone())))
+    }
+}
+
+impl<T> MaybeNot<T> {
+    pub fn into_result(self) -> Result<T, Variant> {
+        self.0
+    }
+
+    pub fn ok(self) -> Option<T> {
+        self.0.ok()
+    }
+}
+
+impl<T: ToVariant, E: ToVariant> ToVariant for Result<T, E> {
+    fn to_variant(&self) -> Variant {
+        let mut dict = Dictionary::new();
+        match &self {
+            Ok(val) => dict.set(&"Ok".into(), &val.to_variant()),
+            Err(err) => dict.set(&"Err".into(), &err.to_variant()),
+        }
+        dict.to_variant()
+    }
+}
+
+impl<T: FromVariant, E: FromVariant> FromVariant for Result<T, E> {
+    fn from_variant(variant: &Variant) -> Option<Self> {
+        let dict = variant.try_to_dictionary()?;
+        if dict.len() != 1 {
+            return None;
+        }
+        let keys = dict.keys();
+        let key_variant = keys.get_ref(0);
+        let key = key_variant.try_to_string()?;
+        match key.as_str() {
+            "Ok" => {
+                let val = T::from_variant(dict.get_ref(key_variant))?;
+                Some(Ok(val))
+            },
+            "Err" => {
+                let err = E::from_variant(dict.get_ref(key_variant))?;
+                Some(Err(err))
+            },
+            _ => None,
+        }
+    }
+}
+
+impl<T: ToVariant> ToVariant for &[T] {
+    fn to_variant(&self) -> Variant {
+        let mut array = VariantArray::new();
+        for val in self.iter() {
+            // there is no real way to avoid CoW allocations right now, as ptrw isn't exposed
+            array.push(&val.to_variant());
+        }
+        array.to_variant()
+    }
+}
+
+impl<T: ToVariant> ToVariant for Vec<T> {
+    fn to_variant(&self) -> Variant {
+        self.as_slice().to_variant()
+    }
+}
+
+impl<T: FromVariant> FromVariant for Vec<T> {
+    fn from_variant(variant: &Variant) -> Option<Self> {
+        use std::convert::TryInto;
+
+        let arr = variant.try_to_array()?;
+        let len: usize = arr.len().try_into().ok()?;
+        let mut vec = Vec::with_capacity(len);
+        for idx in 0..len as i32 {
+            if let Some(item) = T::from_variant(arr.get_ref(idx)) {
```

Comment:
This is a lossy conversion, I don't know what would be the expected behavior honestly.

I would probably expect a heterogeneous array to not be convertible to a homogeneous vector. It would warn you if the caller's code does something weird, it's basically adding run-time checked static typing.

Reply:
You're right about that. It's better to have the user handle values explicitly with `MaybeNot<T>`, if they really want to. The only problem I see is that `godot_wrap_method`'s error message is not very clear about what the expected type is:

```rust
godot_error!("Incorrect argument type {:?} for argument {}",
    _variant.get_type(),
    offset);
```

I'll see if I can improve on that somehow. Maybe `stringify` the argument name and Rust type?

Reply:
Modified the error message to something like: `Incorrect argument type for argument #1 (foo: Vec<Option<Result<i64, u8>>>). Got VariantType::VariantArray (non-primitive types may impose structural checks)`. I think this is enough information.